example/hash.cpp

```diff
@@ -0,0 +1,166 @@
+//
+// Copyright (c) 2020 Richard Hodges (hodges.r@gmail.com)
+//
+// Distributed under the Boost Software License, Version 1.0. (See accompanying
+// file LICENSE_1_0.txt or copy at http://www.boost.org/LICENSE_1_0.txt)
+//
+// Official repository: https://github.com/boostorg/json
+//
+
+//[example_hash
+
+/*
+    This example demonstrates how to provide specializations of
+    boost::hash and std::hash for a json value which provides compatibility
+    with the behaviour of operator== so that json values can be safely stored in
+    an unordered_map. Note that for small maps and large objects, the hashing function
+    will dominate. But for large maps indexed by small json keys this could be useful.
+*/
+
+#include <boost/json.hpp>
+#include <boost/functional/hash.hpp>
+#include <iostream>
+#include <unordered_map>
+
+namespace boost {
+
+// forward declarations
+
+template <>
+struct hash< json::array >
+{
+    inline
+    std::size_t
+    operator()(json::array const &arg) const;
+};
+
+template <>
+struct hash< json::object >
+{
+    inline
+    std::size_t
+    operator()(json::object const &arg) const;
+};
+
+template <>
+struct hash< json::string >
+{
+    inline
+    std::size_t
+    operator()(json::string const &arg) const;
+};
+
+template <>
+struct hash< json::value >
+{
+    inline
+    std::size_t
+    operator()(json::value const &arg) const;
+};
+
+template <>
+struct hash< std::nullptr_t >
+{
+    inline
+    std::size_t
+    operator()(std::nullptr_t) const
+    {
+        return 0;
+    }
+};
+
+std::size_t
+hash< json::value >::operator()(json::value const &arg) const
+{
+    using namespace boost::json;
+    return json::visit([](auto&& item){
+        return hash<std::decay_t<decltype(item)>>()(item);
+    }, arg);
+}
+
+std::size_t hash< json::array >::operator()(json::array const &arg) const
+{
+    std::size_t seed = 0;
+    for (auto &x : arg)
+        hash_combine(seed, x);
+    return seed;
+}
+
+std::size_t hash< json::object >::operator()(json::object const &arg) const
+{
+    std::size_t                                   seed = 0;
+    thread_local std::vector< json::string_view > keys;
```

Comment:
No need for thread_local on a local variable.

Reply:
I think he meant to wrtite `thread_local static`